### PR TITLE
Validate allowed action

### PR DESCRIPTION
**Pull request by mattcollier**

### CODE REVIEW

File: lib/CapabilityDelegation.js [hunk 1]
```diff
@@ -62,6 +67,11 @@ module.exports = class CapabilityDelegation extends ControllerProofPurpose {
       }
     }
     this.suite = suite;
+
+    if(currentDate && !(currentDate instanceof Date)) {
```
Comment (davidlehn) [inline lib/CapabilityDelegation.js]: Docs for this class say date defaults to `new Date()`, but I think that default happens in util.verifyCapabilityChain?  Maybe this check should move into that function?  Now you could pass in null here and it would get through to the comparisons.  Might be able to just do the `instanceof` check.

File: lib/CapabilityInvocation.js [hunk 2]
```diff
@@ -60,13 +66,24 @@ module.exports = class CapabilityInvocation extends ControllerProofPurpose {
       }
     }
     this.suite = suite;
+
+    if(currentDate && !(currentDate instanceof Date)) {
```
Comment (davidlehn) [inline lib/CapabilityInvocation.js]: Same issues as in CapabilityDelegation.

File: lib/utils.js [hunk 3]
```diff
@@ -490,6 +501,12 @@ api.verifyCapabilityChain = async ({
       throw error;
     }
 
+    // if root.expires is undefined then Date.parse returns NaN which is falsey
+    const rootExpirationDate = Date.parse(root.expires);
```
Comment (davidlehn) [inline lib/utils.js]: Is there a concern with bogus data here (and with the other parse())?  Passing in any unknown format will result in NaN.  Would checking that the parse call worked be too strict?  Or maybe just require the date fields are validated before any of this code is called.

File: lib/utils.js [hunk 4]
```diff
@@ -490,6 +501,12 @@ api.verifyCapabilityChain = async ({
...
+    if(rootExpirationDate && currentDate > rootExpirationDate) {
```
Comment (davidlehn) [inline lib/utils.js]: There's an epoch edge case bug here.  Date.parse returns an int.  So root.expires='1970' would result in 0 and skipping the test.  That would make people checking for validity with a currentDate in 1969 sad.  Should perhaps do `new Date(Date.parse(root.expires))` to get a Date object, but then need to deal with the potential 'Invalid Date' result for an invalid root.expires.  Dates are hard.  While this example is unlikely, there may be tricky data that could slip through the checks.

File: lib/CapabilityDelegation.js (hunk 1 above)
Comment (mattcollier) [reply lib/CapabilityDelegation.js]: Attempted to address your concerns here: 32e102b746c3b6443e855917c5a9a51b34b83b9d

File: lib/CapabilityDelegation.js [hunk 5]
```diff
@@ -40,11 +40,15 @@ module.exports = class CapabilityDelegation extends ControllerProofPurpose {
...
+   * @param [currentDate = new Date()] {Date} - The date used for comparison
+   *   when determining if a capability has expired.
+   * @param [maxChainLength=10] - The maximum length of the capability
+   *   delegation chain.
    */
   constructor({
     capabilityChain, verifiedParentCapability,
     expectedTarget, expectedRootCapability, inspectCapabilityChain,
-    capability, capabilityAction, caveat, suite,
+    capability, capabilityAction, caveat, suite, currentDate, maxChainLength,
```
Comment (dlongley) [inline lib/CapabilityDelegation.js]: The docs say default `maxChainLength=10` ... so did we want this? Same thing appears in a few other places.

```suggestion
    capability, capabilityAction, caveat, suite, currentDate, maxChainLength = 10,
```

Hmm, I see it's set to `constants.MAX_CHAIN_LENGTH` elsewhere ... so maybe this is ok at this layer? Also, do we support having the docs say something like `constants.MAX_CHAIN_LENGTH` and then we define that elsewhere -- or should that be used internally only (slight worry about things getting out of sync, maybe not important).

File: lib/utils.js (hunk 4 above)
Comment (mattcollier) [reply lib/utils.js]: @davidlehn do we have code from jsonld-sigs or elsewhere that addresses these concerns?  We check for expired dates like this in other libraries right?

File: lib/CapabilityInvocation.js (hunk 2 above)
Comment (mattcollier) [reply lib/CapabilityInvocation.js]: Attempted to address in 32e102b746c3b6443e855917c5a9a51b34b83b9d

File: lib/CapabilityDelegation.js (hunk 5 above)
Comment (mattcollier) [reply lib/CapabilityDelegation.js]: The concern about sync is what informed the current code.  I don't know what the best solutions is.  Putting `constants.MAX_CHAIN_LENGTH`in the jsdocs is very unhelpful if one would like to know what the value is.  And putting a hard coded value has the potential to get out of sync with the code.

File: lib/utils.js (hunk 4 above)
Comment (mattcollier) [reply lib/utils.js]: @davidlehn does this implementation from jsonld-signatures address your concerns? https://github.com/digitalbazaar/jsonld-signatures/blob/531058bf1af7101bd261f540e4b875f3e5101076/lib/purposes/ProofPurpose.js#L49-L58

File: lib/utils.js (hunk 3 above)
Comment (mattcollier) [reply lib/utils.js]: Attempted to address: 585228e46ef73c99ce7a2c436000ac9a20714e04

File: lib/utils.js (hunk 4 above)
Comment (mattcollier) [reply lib/utils.js]: Attempted to address: 585228e46ef73c99ce7a2c436000ac9a20714e04

Comment (davidlehn) [reply lib/utils.js]: Looks more robust now.  It's a challenge to consider odd edge cases.  I imagine this works for regular use cases.

Comment (mattcollier) [reply lib/utils.js]: Yes, the test suite still passes for regular use cases and I just add a couple more test related to Unix epoch edge case.
